SimulationRuntime/c/simulation/results/simulation_result_mat4.cpp

```diff
@@ -0,0 +1,701 @@
+      cur++;
+    }
+
+  for (int i=0; i < mData->nParametersInteger; i++)
+    if (!mData->integerParameterData[i].filterOutput) {
+      memcpy((uint8_t*)name + maxLengthName * cur, mData->integerParameterData[i].info.name, strlen(mData->integerParameterData[i].info.name));
+      memcpy((uint8_t*)description + maxLengthDesc * cur, mData->integerParameterData[i].info.comment, strlen(mData->integerParameterData[i].info.comment));
+      cur++;
+    }
+
+  for (int i=0; i < mData->nParametersBoolean; i++)
+    if (!mData->booleanParameterData[i].filterOutput) {
+      memcpy((uint8_t*)name + maxLengthName * cur, mData->booleanParameterData[i].info.name, strlen(mData->booleanParameterData[i].info.name));
+      memcpy((uint8_t*)description + maxLengthDesc * cur, mData->booleanParameterData[i].info.comment, strlen(mData->booleanParameterData[i].info.comment));
+      cur++;
+    }
+
+  for (int i=0; i < mData->nAliasReal; i++)
+    if (!mData->realAlias[i].filterOutput) {
+      memcpy((uint8_t*)name + maxLengthName * cur, mData->realAlias[i].info.name, strlen(mData->realAlias[i].info.name));
+      memcpy((uint8_t*)description + maxLengthDesc * cur, mData->realAlias[i].info.comment, strlen(mData->realAlias[i].info.comment));
+      cur++;
+    }
+
+  for (int i=0; i < mData->nAliasInteger; i++)
+    if (!mData->integerAlias[i].filterOutput) {
+      memcpy((uint8_t*)name + maxLengthName * cur, mData->integerAlias[i].info.name, strlen(mData->integerAlias[i].info.name));
+      memcpy((uint8_t*)description + maxLengthDesc * cur, mData->integerAlias[i].info.comment, strlen(mData->integerAlias[i].info.comment));
+      cur++;
+    }
+
+  for (int i=0; i < mData->nAliasBoolean; i++)
+    if (!mData->booleanAlias[i].filterOutput) {
+      memcpy((uint8_t*)name + maxLengthName * cur, mData->booleanAlias[i].info.name, strlen(mData->booleanAlias[i].info.name));
+      memcpy((uint8_t*)description + maxLengthDesc * cur, mData->booleanAlias[i].info.comment, strlen(mData->booleanAlias[i].info.comment));
+      cur++;
+    }
+
+  //       Name: name
+  //       Rank: 2
+  // Dimensions: maxLength x nVars
+  // Class Type: Character Array
+  //  Data Type: 8-bit, unsigned integer
+  writeMatVer4Matrix_4(matData->pFile, "name", maxLengthName, matData->nSignals, name, MatVer4Type_CHAR);
+  free(name);
+  name = NULL;
+
+  //       Name: description
+  //       Rank: 2
+  // Dimensions: maxLength x nVars
+  // Class Type: Character Array
+  //  Data Type: 8-bit, unsigned integer
+  writeMatVer4Matrix_4(matData->pFile, "description", maxLengthDesc, matData->nSignals, description, MatVer4Type_CHAR);
+  free(description);
+  description = NULL;
+  rt_accumulate(SIM_TIMER_OUTPUT);
+}
+
+#define WRITE_REAL_VALUE(data, offset, value) {if (omc_flag[FLAG_SINGLE_PRECISION]) {float f=(value); memcpy(((uint8_t*)(data)) + (offset)*sizeof(float), &f, sizeof(float));} else {double d=(value); memcpy(((uint8_t*)(data)) + (offset)*sizeof(double), &d, sizeof(double));}}
```

Comment:
@sjoelund Is the usage of `memcpy` in this case slow?

Reply:
Or, updating the data2 header information causes the performance issues.

Reply:
You mean the memcpy during initialization for the variable names? It's done a single time so it should be fine. Writing directly to file of course faster since it's not really necessary?

Do you mean you update the number of columns after each write into data_2? If so, it's totally unnecessary. If the simulation crashes, that might not be written anyway and you need to recover the data...

Reply:
No, I was talking about `memcpy` in `WRITE_REAL_VALUE`, but never mind.

It is possible to load partial results while the simulation is running if one updates the header constantly. However, that slows down the entire simulation. It's not really that much, but anyway I guess I will just write the header once at the end of the simulation.

Reply:
It should be faster to write directly instead of memcpy, but I guess only marginally so

Reply:
Regarding the possibility to read the result files during the simulation, IMO this is an essential requirement when you are running simulations that take more than 10 seconds, which are not that uncommon in actual industrial practice. 

If something goes wrong after 1 minute of a 30-minutes long simulation, you don't wanna wait 30 minutes to find out.

I understand @sjoelund is working on this topic for the OpenCPS project.

Maybe it is ok to update the file once every N intervals, and in this case I'd suggest to have a simulation flag to set that N.

Reply:
Or we just have an option to ask the user if we should automatically repair the header if number of columns is 0 (and then deduce how large the matrix is by assuming the rest of the file contains the data_2 matrix)